File: python/bookstore_client.py
import grpc
from google.protobuf import empty_pb2

import bookstore_pb2
import bookstore_pb2_grpc
import hello_pb2
import hello_pb2_grpc
from grpc_interceptor import create_channel, AuthOptions, ServiceAccountAuth


def run(host: str, port: int, *,
        api_key: str = None, service_account_path: str = None, service_name: str = None,
        insecure: bool = False, cacert_path: str = None):
    timeout = 3  # seconds
    auth_options = AuthOptions(
        api_key=api_key,
        service_account_auth=ServiceAccountAuth.of(
            service_account_path=service_account_path,
            service_name=service_name
        ) if service_account_path else None
    )
    cacert = None
    if cacert_path:
        with open(cacert_path, "rb") as f:
            cacert = f.read()
    channel = create_channel(f"{host}:{port}", auth_options=auth_options, insecure=insecure, cacert=cacert)
    stub = bookstore_pb2_grpc.BookstoreStub(channel)

    try:
        shelves = stub.ListShelves(empty_pb2.Empty(), timeout)
        print(f"ListShelves: {shelves}")

        first_shelf = stub.GetShelf(bookstore_pb2.GetShelfRequest(shelf=shelves.shelves[0].id), timeout)
        print(f"GetShelf: {first_shelf}")

        stub.DeleteShelf(bookstore_pb2.DeleteShelfRequest(shelf=-1), timeout)
        print("DeleteShelf")
    except grpc.RpcError as rpc_error:
        print(f"{rpc_error.code()}: {rpc_error.details()}")
        # raise rpc_error
    finally:
        channel.close()


def hello():
    timeout = 3  # seconds
    channel = create_channel("examples.http-client.intellij.net:443")
    stub = hello_pb2_grpc.HelloServiceStub(channel)

    try:
        resp = stub.SayHello(hello_pb2.HelloRequest(greeting="This is greeting"), timeout)
        print(f"SayHello: {resp}")
    except grpc.RpcError as rpc_error:
        print(f"{rpc_error.code()}: {rpc_error.details()}")
        raise rpc_error
    finally:
        channel.close()


if __name__ == "__main__":
    _host = "34.87.33.91"
    _api_key_port = 9000
    _api_key_ssl_port = 443
    _auth_token_port = 9001
    _api_key = None
    _service_account_file = "/Users/datle/experimental-229809-32e56a607af1.json"  # dat-service-account
    _cacert_file = "/Users/datle/Downloads/bookstore.example.com.pem"
    _service_name = "bookstore2.endpoints.experimental-229809.cloud.goog"

    # hello()
    # run(_host, _api_key_port, insecure=True)
    # run(_host, _api_key_ssl_port, api_key=_api_key, cacert_path=_cacert_file)
    run(_host, _auth_token_port, service_account_path=_service_account_file, service_name=_service_name,
        cacert_path=_cacert_file)

File: src/main/java/com/datle/esp/auth/BookstoreClient.java
package com.datle.esp.auth;

import com.datle.esp.auth.AuthInterceptor.AuthOptions;
import com.datle.esp.auth.JwtToken.ServiceAccountAuth;
import com.google.endpoints.examples.bookstore.BookstoreGrpc;
import com.google.endpoints.examples.bookstore.BookstoreGrpc.BookstoreBlockingStub;
import com.google.endpoints.examples.bookstore.DeleteShelfRequest;
import com.google.endpoints.examples.bookstore.GetShelfRequest;
import com.google.endpoints.examples.bookstore.ListShelvesResponse;
import com.google.endpoints.examples.bookstore.Shelf;
import com.google.protobuf.Empty;
import io.grpc.Deadline;
import io.grpc.Grpc;
import io.grpc.StatusRuntimeException;
import io.grpc.TlsChannelCredentials;
import java.io.FileInputStream;
import java.io.IOException;
import java.util.concurrent.TimeUnit;

public class BookstoreClient {

    private static final Deadline TIMEOUT = Deadline.after(3, TimeUnit.SECONDS);

    public static void main(String[] args) throws IOException {
        String host = "34.87.33.91";
        int apiKeyPort = 9000;
        int authTokenPort = 9001;
        String apiKey = null;
        String serviceAccountFilePath = "/Users/datle/experimental-229809-32e56a607af1.json"; // dat-service-account
        String serviceName = "bookstore2.endpoints.experimental-229809.cloud.goog";
        String cacertPath = "/Users/datle/Downloads/bookstore.example.com.pem";

        ServiceAccountAuth serviceAccountAuth = ServiceAccountAuth.fromFile(serviceAccountFilePath, serviceName, 120);
        AuthOptions authOptions = new AuthOptions(apiKey, serviceAccountAuth);

        var creds = TlsChannelCredentials.newBuilder().trustManager(new FileInputStream(cacertPath)).build();
//        creds = InsecureChannelCredentials.create();
        var channel = Grpc.newChannelBuilderForAddress(host, authTokenPort, creds)
                .intercept(new AuthInterceptor(authOptions))
                .build();
        BookstoreBlockingStub stub = BookstoreGrpc.newBlockingStub(channel);

        try {
            ListShelvesResponse listShelvesResponse = stub.withDeadline(TIMEOUT)
                    .listShelves(Empty.getDefaultInstance());
            System.out.printf("listShelves=%s%n", listShelvesResponse);

            Shelf firstShelf = stub.withDeadline(TIMEOUT)
                    .getShelf(GetShelfRequest.newBuilder().setShelf(listShelvesResponse.getShelves(0).getId()).build());
            System.out.printf("getShelf=%s%n", firstShelf);

            stub.withDeadline(TIMEOUT).deleteShelf(DeleteShelfRequest.newBuilder().setShelf(-1).build());
            System.out.printf("deleteShelf%n");
        } catch (StatusRuntimeException e) {
            System.out.println(e.getMessage());
        } finally {
            channel.shutdownNow();
        }
    }
}
